File: src/ralph_wiggum/commands/cancel.py
"""Cancel command - cancel an active Ralph loop."""

from pathlib import Path

from rich.console import Console

from ralph_wiggum.models.state import LoopState

console = Console()

STATE_FILE = Path(".kiro/ralph-loop.local.md")
SESSION_FILE = Path(".kiro/ralph-session.json")


def cancel_cmd() -> None:
    """Cancel an active Ralph loop.

    Removes the loop state file and session file, stopping
    the loop on its next iteration check.
    """
    if not STATE_FILE.exists():
        console.print("No active Ralph loop found.")
        return

    try:
        state = LoopState.from_markdown(STATE_FILE.read_text())
        iteration = state.iteration
    except Exception:
        iteration = "?"

    STATE_FILE.unlink(missing_ok=True)
    SESSION_FILE.unlink(missing_ok=True)

    console.print(f"Cancelled Ralph loop (was at iteration {iteration})")

File: src/commands/cancel.ts
/**
 * @fileoverview Cancel command for Ralph Wiggum CLI.
 * Stops an active loop by removing state files.
 * @module commands/cancel
 */
import { unlink } from "node:fs/promises";
import { log } from "@clack/prompts";

import { type LoopState, stateFromJson } from "../schemas/state";
import { SESSION_FILE, STATE_FILE } from "../utils/paths";

/**
 * Cancels an active Ralph Wiggum loop.
 * Marks the state as inactive so it can be resumed later.
 * @returns Resolves when cancellation is complete
 */
export async function cancelCommand(): Promise<void> {
	const stateFile = Bun.file(STATE_FILE);

	if (!(await stateFile.exists())) {
		log.message("No active Ralph loop found.");
		return;
	}

	// Read and parse the existing state
	let state: LoopState;
	let iteration: number | string = "?";
	try {
		const content = await stateFile.text();
		state = stateFromJson(content);
		iteration = state.iteration;
	} catch {
		// If we can't parse the state, just delete it
		await unlink(STATE_FILE).catch(() => {});
		log.message("Cancelled Ralph loop (invalid state file)");
		return;
	}

	// Mark state as inactive (preserves for resume)
	state.active = false;

	// Write updated state back
	await Bun.write(STATE_FILE, JSON.stringify(state));

	// Delete session file (not needed for resume)
	await unlink(SESSION_FILE).catch(() => {});

	log.message(`Cancelled Ralph loop at iteration ${iteration}`);
	log.message("Run 'ralph resume' to continue where you left off.");
}
